tools/autograd/templates/Functions.cpp

```diff
@@ -165,6 +165,31 @@ Tensor mm_mat2_backward(const Tensor & grad, const Tensor & mat1, IntList sizes,
   }
 }
 
+Tensor renorm_backward(const Tensor & grad, const Tensor & self, Scalar p, int64_t dim, Scalar maxnorm) {
+  auto transposed_sizes = std::vector<int64_t>(self.transpose(dim, 0).sizes());
+  auto flatten = [&](const Tensor & t) {
+    return t.transpose(dim, 0).contiguous().view({t.size(dim), -1});
+  };
+  auto unflatten = [&](const Tensor & t) {
+    return t.contiguous().view(transposed_sizes).transpose(dim, 0);
+  };
+
+  // renorm computes the norm over all dimensions except `dim`, which is why
+  // we need the flatten and unflatten business. TODO: simplify this when we
+  // add support for norm over multiple dimensions.
+  auto self_flat = flatten(self);
+  auto grad_flat = flatten(grad);
+  auto norm_flat = self_flat.norm(p, 1, true);
+  auto grad_output = (self_flat * grad_flat).sum(1, true);
+  auto nb = norm_backward(grad_output, self_flat, p, norm_flat, 1, true);
+  auto invnorm = (norm_flat + 1e-7).reciprocal();
+  auto grad_norm = unflatten(maxnorm * invnorm * (grad_flat - invnorm * nb));
+  auto norm = unflatten(norm_flat.expand_as(self_flat));
+
+  auto mask = Variable(norm < maxnorm).detach();
```

Comment:
Why the `detach` here?

Reply:
This is a little weird. Without this, we end up trying to negate a `ByteTensor` in the double-backwards. `ByteTensor` is unsigned so the negation fails.

The problem is because the comparison `mask = norm < maxnorm` currently has the result have `requires_grad=True`. Later on we do `1 - mask`, which is where the negation in the double-backwards comes in.

Reply:
I think adding something like TensorFlows `tf.where` would help here. We want a way to express element-wise `cond ? a : b` in Tensor operations. The current method of multiplications and `1 - cond` isn't ideal.

Reply:
Why does comparison has `requires_grad=True`?

Reply:
But yes, `np.where`-like op would be useful.

Reply:
I have an issue tracking it here:
https://github.com/pytorch/pytorch/issues/3348

We should fix it, but I think we'll also need to change some tests. IIRC, a bunch of them will fail because the backward engine throws an exception if things don't require grad.